module/zfs/arc.c

```diff
@@ -8685,6 +8685,27 @@ l2arc_stop(void)
 	mutex_exit(&l2arc_feed_thr_lock);
 }
 
+#ifdef _KERNEL
+static int
+param_set_arc_value(const char *buf, zfs_kernel_param_t *kp)
+{
+	uint64_t val;
+	int error;
+
+	error = kstrtoull(buf, 0, &val);
+	if (error)
+		return (SET_ERROR(error));
+
+	error = param_set_long(buf, kp);
+	if (error < 0)
+		return (SET_ERROR(error));
+
+	arc_tuning_update();
+
+	return (0);
+}
+#endif
```

Comment:
This _may_ need to be moved to `./module/os/linux/zfs/arc_os.c` unless we declare portable wrappers to do this.

Reply:
I am still a bit lost due to all these changes to the codebase, i am doing this locally to see what happens.

Reply:
Moved to `./module/os/linux/zfs/arc_os.c`, everything seems to be broken now in `./module/zfs/arc.c`:

```
  CC [M]  /tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/module/spl/../os/linux/spl/spl-zlib.o
  LD [M]  /tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/module/spl/spl.o
  CC [M]  /tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/module/zfs/aggsum.o
  CC [M]  /tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/module/zcommon/zfs_prop.o
  CC [M]  /tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/module/zfs/arc.o
In file included from include/linux/module.h:17:0,
                 from /tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/include/os/linux/spl/sys/atomic.h:28,
                 from /tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/include/sys/zfs_context.h:35,
                 from /tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/include/sys/spa.h:37,
                 from /tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/module/zfs/arc.c:278:
/tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/module/zfs/arc.c:8699:47: error: 'param_set_arc_long' undeclared here (not in a function)
 ZFS_MODULE_PARAM_CALL(zfs_arc, zfs_arc_, min, param_set_arc_long,
                                               ^
include/linux/moduleparam.h:190:14: note: in definition of macro 'module_param_call'
    { (void *)set, (void *)get };    \
              ^
/tmp/zfs-build-buildbot-czCOc8ZC/BUILD/zfs-0.8.0/module/zfs/arc.c:8699:1: note: in expansion of macro 'ZFS_MODULE_PARAM_CALL'
 ZFS_MODULE_PARAM_CALL(zfs_arc, zfs_arc_, min, param_set_arc_long,
```

I tried moving more code to `./module/os/linux/zfs/arc_os.c` but made things even worse: i never thought i would miss `#ifdef <platform>`.

Reply:
@loli10K I sympathize!  Refactoring the code to accommodate other platforms is going to take a little adjusting too.  Plus we're still working on getting all the bits in place and documented, and a lot of the details are still being worked through.  So things may change.

It looks to me like the only thing missing from this change is to add the common prototypes to the private `arc_impl.h` header.  The `arc_os.c` sources contain platform specific versions of functions as needed, but when possible we want to share prototypes.  @mattmacy would have the best idea what the `param_set_arc_long` and `param_set_arc_int` counterparts will need to look like for FreeBSD.

This change resolves the build issues for me.  It:

1) Adds the needed forward references to the private `arc_impl.h` header, and
2) Removes the `kstrtoull` and `kstrtoul` calls from the helpers.  We should be able to rely on the kernel helpers to perform the needed checks.  Though we can add them back as long as we cast the type, if there's a case here which isn't covered.


```diff
diff --git a/include/sys/arc_impl.h b/include/sys/arc_impl.h
index 28d430f..8eeee54 100644
--- a/include/sys/arc_impl.h
+++ b/include/sys/arc_impl.h
@@ -611,6 +611,10 @@ extern void arc_prune_async(int64_t);
 extern int arc_memory_throttle(spa_t *spa, uint64_t reserve, uint64_t txg);
 extern uint64_t arc_free_memory(void);
 extern int64_t arc_available_memory(void);
+extern void arc_tuning_update(void);
+
+extern int param_set_arc_long(const char *buf, zfs_kernel_param_t *kp);
+extern int param_set_arc_int(const char *buf, zfs_kernel_param_t *kp);
 
 #ifdef __cplusplus
 }
diff --git a/include/sys/zfs_context.h b/include/sys/zfs_context.h
index 9780726..4ea3050 100644
--- a/include/sys/zfs_context.h
+++ b/include/sys/zfs_context.h
@@ -204,6 +204,10 @@ extern int aok;
 /*
  * Tunables.
  */
+typedef struct zfs_kernel_param {
+       const char *name;       /* unused stub */
+} zfs_kernel_param_t;
+
 #define        ZFS_MODULE_PARAM(scope_prefix, name_prefix, name, type, perm, desc)
 #define        ZFS_MODULE_PARAM_CALL(scope_prefix, name_prefix, name, setfunc, \
        getfunc, perm, desc)
diff --git a/module/os/linux/zfs/arc_os.c b/module/os/linux/zfs/arc_os.c
index 25ce9dc..eecc9fc 100644
--- a/module/os/linux/zfs/arc_os.c
+++ b/module/os/linux/zfs/arc_os.c
@@ -358,16 +358,11 @@ arc_lowmem_fini(void)
        spl_unregister_shrinker(&arc_shrinker);
 }
 
-static int
+int
 param_set_arc_long(const char *buf, zfs_kernel_param_t *kp)
 {
-        uint64_t val;
         int error;
 
-        error = kstrtoull(buf, 0, &val);
-        if (error)
-                return (SET_ERROR(error));
-
         error = param_set_long(buf, kp);
         if (error < 0)
                 return (SET_ERROR(error));
@@ -377,16 +372,11 @@ param_set_arc_long(const char *buf, zfs_kernel_param_t *kp)
         return (0);
 }
 
-static int
+int
 param_set_arc_int(const char *buf, zfs_kernel_param_t *kp)
 {
-        int64_t val;
         int error;
 
-        error = kstrtoul(buf, 0, &val);
-        if (error)
-                return (SET_ERROR(error));
-
         error = param_set_int(buf, kp);
         if (error < 0)
                 return (SET_ERROR(error));
diff --git a/module/zfs/arc.c b/module/zfs/arc.c
index 9bf3f18..d4651d8 100644
--- a/module/zfs/arc.c
+++ b/module/zfs/arc.c
@@ -814,7 +814,6 @@ static void arc_hdr_alloc_abd(arc_buf_hdr_t *, boolean_t);
 static void arc_access(arc_buf_hdr_t *, kmutex_t *);
 static boolean_t arc_is_overflowing(void);
 static void arc_buf_watch(arc_buf_t *);
-static void arc_tuning_update(void);
 
 static arc_buf_contents_t arc_buf_type(arc_buf_hdr_t *);
 static uint32_t arc_bufc_to_flags(arc_buf_contents_t);
@@ -6878,7 +6877,7 @@ arc_state_multilist_index_func(multilist_t *ml, void *obj)
  * updated manually.  Non-zero zfs_* values which differ from the currently set
  * values will be applied.
  */
-static void
+void
 arc_tuning_update(void)
 {
        uint64_t allmem = arc_all_memory();
```

Reply:
Thank you